File: app/src/main/java/com/learning/ruslan/Support.java
package com.learning.ruslan;

import android.content.ContentValues;
import android.content.Context;
import android.content.res.ColorStateList;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.ColorDrawable;
import android.os.Build;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;
import android.text.style.TypefaceSpan;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.Window;

import androidx.annotation.IdRes;
import androidx.annotation.IntRange;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.ActionBar;

import com.learning.ruslan.databases.RusBaseHelper;
import com.learning.ruslan.databases.RusDbSchema.SupportTable;
import com.learning.ruslan.databases.SupportCursorWrapper;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

public class Support {

    //theme constants
    public static final String THEME_LIGHT = "light";
    public static final String THEME_NIGHT = "night";

    //constants for time in learning fragment
    public static final int MIN_TIME = 100;
    public static final int MAX_TIME = 10000;

    //languages
    public static final String RUSSIAN = "rus";
    public static final String ENGLISH = "eng";

    public static final String[] languages = new String[] {RUSSIAN, ENGLISH};

    //colors
    public static final int color_magenta2 = Color.rgb(255, 0, 170);
    public static final int color_red_night = Color.rgb(247, 124, 124);
    public static final int color_red_light = Color.rgb(243, 10, 10);
    public static final int score_color_light = Color.rgb(34, 140, 89);
    public static final int score_color_night = Color.rgb(75, 242, 161);

    private static final Set<String> THEMES =
            Collections.unmodifiableSet(new HashSet<>(Arrays.asList(THEME_LIGHT, THEME_NIGHT)));

    private static Support sSupport;
    private final SQLiteDatabase mDatabase;
    private Settings mSettings;
    private final Context mContext;



    private void setLightScreen(Window window, @Nullable ActionBar actionBar) {
        window.getDecorView().setBackgroundColor(Color.WHITE);
        window.setStatusBarColor(Color.WHITE);
        window.getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);
        if (actionBar != null)
            actionBar.setBackgroundDrawable(new ColorDrawable(Color.WHITE));
    }

    private void setDarkScreen(Window window, @Nullable ActionBar actionBar) {
        window.getDecorView().setBackgroundColor(Color.BLACK);
        window.setStatusBarColor(Color.BLACK);
        window.getDecorView().setSystemUiVisibility(0);
        if (actionBar != null)
            actionBar.setBackgroundDrawable(new ColorDrawable(Color.BLACK));
    }


    @RequiresApi(api = Build.VERSION_CODES.O)
    public void setThemeToScreen(
            Window window,
            @Nullable ActionBar actionBar,
            Menu menu,
            @NonNull @IdRes int[] menuItems)
    {

        int iconColor;
        SpannableString titleText;

        if (getTheme().equals(THEME_LIGHT)) {
            setLightScreen(window, actionBar);
            iconColor = Color.BLACK;
        }
        else {
            setDarkScreen(window, actionBar);
            iconColor = Color.WHITE;
        }


        titleText = new SpannableString(Objects.requireNonNull(actionBar).getTitle());

        titleText.setSpan(
                new ForegroundColorSpan(iconColor),
                0,
                titleText.length(),
                SpannableString.SPAN_EXCLUSIVE_EXCLUSIVE
        );

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            titleText.setSpan(
                    new TypefaceSpan(Typeface.DEFAULT_BOLD),
                    0,
                    titleText.length(),
                    SpannableString.SPAN_EXCLUSIVE_EXCLUSIVE
            );
        }

        titleText.setSpan(
                new TypefaceSpan("font/verdana.ttf"),
                0,
                titleText.length(),
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        );

        actionBar.setTitle(titleText);
//        if (isUpArrow) {
//
//            Drawable upArrow = ContextCompat.getDrawable(mContext, R.drawable.ic_menu_up);
//            Objects.requireNonNull(upArrow).setColorFilter(iconColor, PorterDuff.Mode.SRC_ATOP);
//
//            actionBar.setHomeAsUpIndicator(upArrow);
//        }

        for (int id : menuItems) {
            MenuItem item = menu.findItem(id);

            SpannableString itemTitle = new SpannableString(item.getTitle());
            itemTitle.setSpan(
                    new ForegroundColorSpan(iconColor),
                    0,
                    itemTitle.length(),
                    SpannableString.SPAN_EXCLUSIVE_EXCLUSIVE
            );

            item.setTitle(itemTitle);
            item.setIconTintList(ColorStateList.valueOf(iconColor));
        }
    }





    public static Support get(Context context) {
        if (sSupport == null)
            sSupport = new Support(context);

        return sSupport;
    }

    private Support(Context context) {
        this.mContext = context.getApplicationContext();
        this.mDatabase = new RusBaseHelper(mContext).getWritableDatabase();
        mSettings = new Settings();

        SupportCursorWrapper cursor = querySupports(null, null);

        if (cursor.getCount() == 0)
            createSupportTable();

        cursor.moveToFirst();
        mSettings = cursor.getSettings();
        cursor.close();
    }

    public void setChecked(boolean isChecked) {
        this.mSettings.setChecked(isChecked);
        updateSupport(mSettings);
    }

    public boolean getChecked() {
        return mSettings.isChecked();
    }

    public void setPause(
            @IntRange(from = MIN_TIME, to = MAX_TIME) int pause) {
        this.mSettings.setPause(pause);
        updateSupport(mSettings);
    }

    public int getPause() {
        return mSettings.getPause();
    }

    public void setTheme(String theme) {
        if (THEMES.contains(theme)) {
            this.mSettings.setTheme(theme);
            updateSupport(mSettings);
        }
    }

    public String getTheme() {
        return mSettings.getTheme();
    }

    public void setQuestions(int questions) {
        if (questions < 10) return;
        this.mSettings.setQuestions(questions);
        updateSupport(mSettings);
    }

    public int getQuestions() {
        return mSettings.getQuestions();
    }

    public void setLanguage(String language) {
        this.mSettings.setLanguage(language);
        updateSupport(mSettings);
    }

    public String getLanguage() {
        return mSettings.getLanguage();
    }



    private ContentValues getContentValues(Settings settings) {
        ContentValues values = new ContentValues();
        values.put(SupportTable.Cols.CHECKED, settings.isChecked() ? 1 : 0);
        values.put(SupportTable.Cols.PAUSE, settings.getPause());
        values.put(SupportTable.Cols.THEME, settings.getTheme());
        values.put(SupportTable.Cols.QUESTIONS, settings.getQuestions());
        values.put(SupportTable.Cols.LANGUAGE, settings.getLanguage());

        return values;
    }

    private void createSupportTable() {
        ContentValues values = getContentValues(mSettings);
        mDatabase.insert(SupportTable.NAME, null, values);
    }

    private void updateSupport(Settings settings) {
        String stringId = String.valueOf(settings.getId());
        ContentValues values = getContentValues(settings);

        mDatabase.update(
                SupportTable.NAME,
                values,
                SupportTable.Cols.ID + " = ?",
                new String[] { stringId });
    }

    private SupportCursorWrapper querySupports(String whereClause, String[] whereArgs) {

        Cursor cursor = mDatabase.query(
                SupportTable.NAME,
                null,
                whereClause,
                whereArgs,
                null,
                null,
                null
        );

        return new SupportCursorWrapper(cursor);
    }
}

File: app/src/main/java/com/learning/ruslan/Support.kt
package com.learning.ruslan
//
//import android.content.ContentValues
//import android.content.Context
//import android.content.res.ColorStateList
//import android.database.sqlite.SQLiteDatabase
//import android.graphics.Color
//import android.graphics.Typeface
//import android.graphics.drawable.ColorDrawable
//import android.os.Build
//import android.text.SpannableString
//import android.text.Spanned
//import android.text.style.ForegroundColorSpan
//import android.text.style.TypefaceSpan
//import android.view.*
//import androidx.annotation.IdRes
//import androidx.annotation.RequiresApi
//import androidx.appcompat.app.ActionBar
//import com.learning.ruslan.settings.Settings
//import java.util.*
//
//class Support private constructor(context: Context) {
//    private val mDatabase: SQLiteDatabase
//    private var mSettings: Settings
//    private val mContext = context.applicationContext
//
//    private fun setLightScreen(window: Window, actionBar: ActionBar?) {
//        window.decorView.setBackgroundColor(Color.WHITE)
//        window.statusBarColor = Color.WHITE
//        window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
//        actionBar?.setBackgroundDrawable(ColorDrawable(Color.WHITE))
//    }
//
//    private fun setDarkScreen(window: Window, actionBar: ActionBar?) {
//        window.decorView.setBackgroundColor(Color.BLACK)
//        window.statusBarColor = Color.BLACK
//        window.decorView.systemUiVisibility = 0
//        actionBar?.setBackgroundDrawable(ColorDrawable(Color.BLACK))
//    }
//
//    @RequiresApi(api = Build.VERSION_CODES.O)
//    fun setThemeToScreen(
//        window: Window,
//        actionBar: ActionBar?,
//        menu: Menu?,
//        @IdRes menuItems: IntArray?
//    ) {
//        val iconColor = if (theme == THEME_LIGHT) {
//            setLightScreen(window, actionBar)
//            Color.BLACK
//        } else {
//            setDarkScreen(window, actionBar)
//            Color.WHITE
//        }
//        val titleText = SpannableString(Objects.requireNonNull(actionBar).title)
//        titleText.setSpan(
//            ForegroundColorSpan(iconColor),
//            0,
//            titleText.length,
//            SpannableString.SPAN_EXCLUSIVE_EXCLUSIVE
//        )
//        titleText.setSpan(
//            TypefaceSpan(Typeface.DEFAULT_BOLD),
//            0,
//            titleText.length,
//            SpannableString.SPAN_EXCLUSIVE_EXCLUSIVE
//        )
//        titleText.setSpan(
//            TypefaceSpan("font/verdana.ttf"),
//            0,
//            titleText.length,
//            Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
//        )
//        actionBar?.title = titleText
//        //        if (isUpArrow) {
////
////            Drawable upArrow = ContextCompat.getDrawable(mContext, R.drawable.ic_menu_up);
////            Objects.requireNonNull(upArrow).setColorFilter(iconColor, PorterDuff.Mode.SRC_ATOP);
////
////            actionBar.setHomeAsUpIndicator(upArrow);
////        }
//        if (menu == null || menuItems == null) return
//        for (id in menuItems) {
//            val item = menu.findItem(id)
//            val itemTitle = SpannableString(item.title)
//            itemTitle.setSpan(
//                ForegroundColorSpan(iconColor),
//                0,
//                itemTitle.length,
//                SpannableString.SPAN_EXCLUSIVE_EXCLUSIVE
//            )
//            item.title = itemTitle
//            item.iconTintList = ColorStateList.valueOf(iconColor)
//        }
//    }
//
//    var checked: Boolean
//        get() = mSettings.isChecked
//        set(isChecked) {
//            mSettings.isChecked = isChecked
//            updateSupport(mSettings)
//        }
//    var pause: Int
//        get() = mSettings.pause
//        set(pause) {
//            mSettings.pause = pause
//            updateSupport(mSettings)
//        }
//
//    fun setTheme(theme: String) {
//        if (THEMES.contains(theme)) {
//            mSettings.themeString = theme
//            updateSupport(mSettings)
//        }
//    }
//
//    val theme: String
//        get() = mSettings.themeString
//
//    fun updateTheme() {
//        if (theme == THEME_LIGHT) setTheme(THEME_NIGHT) else setTheme(THEME_LIGHT)
//    }
//
//    var questions: Int
//        get() = mSettings.questions
//        set(questions) {
//            if (questions < 10) return
//            mSettings.questions = questions
//            updateSupport(mSettings)
//        }
//
//    fun setLanguage(position: Int) {
//        mSettings.languageString = languages[position]
//        updateSupport(mSettings)
//    }
//
//    private fun getContentValues(settings: Settings): ContentValues {
//        val values = ContentValues()
//        values.put(SupportTable.Cols.CHECKED, if (settings.isChecked) 1 else 0)
//        values.put(SupportTable.Cols.PAUSE, settings.pause)
//        values.put(SupportTable.Cols.THEME, settings.themeString)
//        values.put(SupportTable.Cols.QUESTIONS, settings.questions)
//        values.put(SupportTable.Cols.LANGUAGE, settings.languageString)
//        return values
//    }
//
//    private fun createSupportTable() {
//        val values = getContentValues(mSettings)
//        mDatabase.insert(SupportTable.NAME, null, values)
//    }
//
//    private fun updateSupport(settings: Settings) {
//        val stringId = settings.id.toString()
//        val values = getContentValues(settings)
//        mDatabase.update(
//            SupportTable.NAME,
//            values,
//            SupportTable.Cols.ID + " = ?", arrayOf(stringId))
//    }
//
//    companion object {
//
//        //colors
//        val color_magenta2 = Color.rgb(255, 0, 170)
//        val color_red_night = Color.rgb(247, 124, 124)
//        val color_red_light = Color.rgb(243, 10, 10)
//        val score_color_light = Color.rgb(34, 140, 89)
//        val score_color_night = Color.rgb(75, 242, 161)
//
//    }
//
//    init {
//        mDatabase = RusBaseHelper(mContext).writableDatabase
//        mSettings = Settings()
//        val cursor = querySupports(null, null)
//        if (cursor.count == 0) createSupportTable()
//        cursor.moveToFirst()
//        mSettings = cursor.settings
//        cursor.close()
//    }
//}